alibi_detect/cd/pytorch/lsdd.py

```diff
@@ -0,0 +1,169 @@
+        preprocess_x_ref
+            Whether to already preprocess and store the reference data.
+        update_x_ref
+            Reference data can optionally be updated to the last n instances seen by the detector
+            or via reservoir sampling with size n. For the former, the parameter equals {'last': n} while
+            for reservoir sampling {'reservoir_sampling': n} is passed.
+        preprocess_fn
+            Function to preprocess the data before computing the data drift metrics.
+        sigma
+            Optionally set the bandwidth of the Gaussian kernel used in estimating the LSDD. Can also pass multiple
+            bandwidth values as an array. The kernel evaluation is then averaged over those bandwidths. If `sigma`
+            is not specified, the 'median heuristic' is adopted whereby `sigma` is set as the median pairwise distance
+            between reference samples.
+        n_permutations
+            Number of permutations used in the permutation test.
+        n_kernel_centers
+            The number of reference samples to use as centers in the Gaussian kernel model used to estimate LSDD.
+            Defaults to 1/20th of the reference data.
+        lambda_rd_max
+            The maximum relative difference between two estimates of LSDD that the regularization parameter
+            lambda is allowed to cause. Defaults to 0.2 as in the paper.
+        device
+            Device type used. The default None tries to use the GPU and falls back on CPU if needed.
+            Can be specified by passing either 'cuda', 'gpu' or 'cpu'.
+        input_shape
+            Shape of input data.
+        data_type
+            Optionally specify the data type (tabular, image or time-series). Added to metadata.
+        """
+        super().__init__(
+            x_ref=x_ref,
+            p_val=p_val,
+            preprocess_x_ref=preprocess_x_ref,
+            update_x_ref=update_x_ref,
+            preprocess_fn=preprocess_fn,
+            sigma=sigma,
+            n_permutations=n_permutations,
+            n_kernel_centers=n_kernel_centers,
+            lambda_rd_max=lambda_rd_max,
+            input_shape=input_shape,
+            data_type=data_type
+        )
+        self.meta.update({'backend': 'pytorch'})
+
+        # set backend
+        if device is None or device.lower() in ['gpu', 'cuda']:
+            self.device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+            if self.device.type == 'cpu':
+                print('No GPU detected, fall back on CPU.')
+        else:
+            self.device = torch.device('cpu')
+
+        if self.preprocess_x_ref or self.preprocess_fn is None:
+            x_ref = torch.as_tensor(self.x_ref)
+            self._configure_normalization(x_ref)
+            x_ref = self._normalize(x_ref)
+            self._initialize_kernel(x_ref)
+            self._configure_kernel_centers(x_ref)
+            self.x_ref = x_ref.numpy()
```

Comment:
`self.x_ref` includes the kernel centers, is this correct?

Reply:
I think so. It's stored as a property for the user to refer to if desired, but is not used by the `score()` method which, when `x_ref` is preprocessed, accesses it through `self.k_xc` only.